src/vivarium/framework/results/strategies.py

```diff
@@ -0,0 +1,377 @@
+        self._pool = defaultdict(lambda: defaultdict(list))
+
+    def add_producer_strategy(self, measure: str, data_filter: str,
+                              groupby_columns: List[str],  aggregator: Callable[[pd.DataFrame], float],
+                              when: str, **additional_keys: str):
+        """Add a new producer strategy to the pool.
+
+        Parameters
+        ----------
+        measure
+        data_filter
+        groupby_columns
+        aggregator
+        when
+        additional_keys
+
+        Returns
+        -------
+
+        """
+        grouper = Grouper(data_filter, groupby_columns)
+        producer = ResultProducerStrategy(measure, aggregator, additional_keys)
+        self._pool[when][grouper].append(producer)
+
+    def produce_results(self, when: str, data: pd.DataFrame) -> Iterator[Result]:
+        """Generate results from data according to strategies in the pool.
+
+        Parameters
+        ----------
+        when
+        data
+
+        Yields
+        ------
+
+        """
+        for grouper, producers in self._pool[when]:
+            grouped_data = grouper.group(data)
+            for producer in producers:
+                result = producer(grouped_data)
+                result.data = grouper.ungroup(result.data)
+                yield result
+
+
+class FormattingStrategy(abc.ABC):
+    """Base interface for results formatting strategies.
+
+    Formatting strategies turn aggregated results into a final output
+    format for results production.
+
+    """
+
+    @staticmethod  # NOTE: Do not override.  This method is final.
+    def _broadcast_aggregates(aggregate_data: pd.Series) -> pd.DataFrame:
+        """Broadcasts aggregate data over unobserved categories.
+
+        This method expects that data provided to the formatter call
+        is a series with a categorical index or a multi-index whose
+        levels are categorical.  The purpose is to ensure that the
+        schema of the final results is consistent by assigning values
+        to all possible observations.
+
+        Parameters
+        ----------
+        aggregate_data
+            Data that has been aggregated to a series of float values.
+
+        Returns
+        -------
+            A dataframe with zero fills for all unobserved groups in the
+            provided series index.
+
+        """
+        if isinstance(aggregate_data.index, pd.MultiIndex):  # Multiple stratification criteria
+            full_index = pd.MultiIndex.from_product(aggregate_data.index.levels,
+                                                    names=aggregate_data.index.names)
+            data = pd.Series(data=0, index=full_index, name=aggregate_data.name)
+            data.loc[aggregate_data.index] = aggregate_data
+            data = data.reset_index()
+        elif isinstance(aggregate_data.index, pd.CategoricalIndex):  # Single stratification criteria
+            full_index = pd.CategoricalIndex(aggregate_data.index.categories,
```

Comment:
Why do we need to build a new index? How is `full_index` different from `aggregate_data.index`?